devdoc/channel.md

```diff
@@ -60,11 +60,23 @@ The value of the `CHANNEL_CALLBACK_RESULT` stored in the `operation` can only ch
 
 `channel_pull` and `channel_push` can be called from callbacks of this module. Since the callbacks are executed on threadpool threads, there is no risk of stack overflow or deadlock.
 
-## Internal Reference Counting
+## Need for `CHANNEL_INTERNAL`
 
-`THANDLE(CHANNEL)` maintains a reference to an internal `THANDLE(CHANNEL_INTERNAL)`. `CHANNEL_INTERNAL` is the object that contains the state of the module. This two-level reference counting is needed because the lifetime of the module is linked to the `THANDLE(CHANNEL)` and the `THANDLE(ASYNC_OP)`s returned by `channel_pull` and `channel_push`. 
+The lifetime of the state of the `channel` module is linked to two things:
+ - `THANDLE(CHANNEL)` references held by the producer and consumer.
+ - `THANDLE(ASYNC_OP)` operations returned from `channel_push` and `channel_pull`. The `THANDLE(ASYNC_OP)`s need to hold references to the state of the `channel` because `async_op_cancel` needs to be able to lock the list of pending operations to remove the operation from the list.
+
+Since the `THANDLE(ASYNC_OP)`s can outlive the `THANDLE(CHANNEL)` references and `THANDLE(ASYNC_OP)` must access the state of the `channel`, the state of the `channel` cannot be freed when the last `THANDLE(CHANNEL)` reference is released. 
+
+To decouple the lifetime of the state from the `THANDLE(CHANNEL)` references, `THANDLE(CHANNEL_INTERNAL)` is used to store the state of the module. `THANDLE(CHANNEL)` holds a single reference to `THANDLE(CHANNEL_INTERNAL)` and each `THANDLE(ASYNC_OP)` also holds a reference to `THANDLE(CHANNEL_INTERNAL)`. When the last `THANDLE(CHANNEL)` reference is released, all pending operations are abandoned and `THANDLE(CHANNEL)` releases its reference to `THANDLE(CHANNEL_INTERNAL)`. However, `THANDLE(CHANNEL_INTERNAL)` is not freed yet because `THANDLE(ASYNC_OP)`s still hold references to it.
+
+## Future Work
+
+`CHANNEL_INTERNAL` is only needed because `THANDLE` does not support a two-step dispose (close and then destroy). If `THANDLE` were to support two kinds of references: an `open-close` type of reference and a `create-destroy` kind of reference, then `CHANNEL_INTERNAL` would not be needed.
 
```

Comment:
I don't think that THANDLE should have that, so I would say we should remove this not to confuse readers.

Reply:
Removed